File: app/src/main/java/me/patrick/aikodigital/pontocerto/network/ServiceProvider.java
package me.patrick.aikodigital.pontocerto.network;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ServiceProvider {

    private static final String API_URL = "https://aiko-olhovivo-proxy.aikodigital.io";
    private static final String API_TOKEN = "24598693b866d61affd4379027baaf0f4d7e09dc77127c6332c553504e00c24f";

    private static Retrofit retrofit;

    private static OkHttpClient.Builder httpClientBuilder = new OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS);

    private static OkHttpClient getHttpClient(final String token) {
        if (token != null) {
            httpClientBuilder.addInterceptor(new Interceptor() {
                @Override
                public Response intercept(Chain chain) throws IOException {
                    Request original = chain.request();
                    Request.Builder requestBuilder = original.newBuilder()
                            .header("Authorization", token);
                    Request request = requestBuilder.build();
                    return chain.proceed(request);
                }
            });
        }
        return httpClientBuilder.build();
    }

    public static Retrofit getRetrofitInstance(String token) {
        if (retrofit == null) {
            retrofit = new Retrofit.Builder()
                    .baseUrl(API_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .client(getHttpClient(token))
                    .build();
        }
        return retrofit;
    }

    public static <T> T createService(Class<T> serviceClass) {
        return getRetrofitInstance(API_TOKEN).create(serviceClass);
    }
}

File: app/app/src/main/java/com/example/app/network/ServiceProvider.kt
package com.example.app.network

import com.example.app.BuildConfig
import com.example.app.data.api.ServiceAPI
import com.example.app.util.BASE_URL
import kotlinx.coroutines.runBlocking
import okhttp3.*
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

class ServiceProvider {
    private var apiCredentials: String = ""

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .addInterceptor(HttpLoggingInterceptor().apply {
            level = HttpLoggingInterceptor.Level.BODY
        })
        .authenticator { _, response ->
            synchronized(this) {
                runBlocking {
                    val isAuthenticated = authenticate(BuildConfig.OLHO_VIVO_API_KEY)
                    if (isAuthenticated) {
                        response.request.newBuilder()
                            .header("Cookie", apiCredentials)
                            .build()
                    } else {
                        null
                    }
                }
            }
        }
        .build()

    private var retrofit: Retrofit = Retrofit.Builder()
        .baseUrl(BASE_URL)
        .addConverterFactory(GsonConverterFactory.create())
        .client(client)
        .build()

    fun <API> createService(apiClass: Class<API>): API = retrofit.create(apiClass)

    private suspend fun authenticate(token: String): Boolean {
        val apiService = createService(ServiceAPI::class.java)
        return try {
            val response = apiService.authenticate(token)
            if (response.isSuccessful) {
                val headers = response.headers()
                val cookies = headers.values("Set-Cookie")
                apiCredentials = cookies.joinToString(", ")
                true
            } else {
                false
            }
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }
}
